File: lb2/calculator/src/main/java/org/example/Calculator.java
package org.example;

import java.math.BigDecimal;

public class Calculator {

    private final ILogger logger;

    public Calculator(ILogger logger) {
        this.logger = logger;
    }

    public BigDecimal add(BigDecimal a, BigDecimal b) {
        BigDecimal result = a.add(b);
        logger.log(a + " + " + b + " = " + result);
        return result;
    }

    public BigDecimal subtract(BigDecimal a, BigDecimal b) {
        BigDecimal result = a.subtract(b);
        logger.log(a + " - " + b + " = " + result);
        return result;
    }

    public BigDecimal multiply(BigDecimal a, BigDecimal b) {
        BigDecimal result = a.multiply(b);
        logger.log(a + " * " + b + " = " + result);
        return result;
    }

    public BigDecimal divide(BigDecimal a, BigDecimal b) {

        if (b.compareTo(BigDecimal.ZERO) == 0) {
            throw new ArithmeticException("Division by zero");
        }

        BigDecimal result = a.divide(b);
        logger.log(a + " / " + b + " = " + result);

        return result;
    }
}

File: pz1/app/src/main/java/com/example/calculator/Calculator.kt
package com.example.calculator

class Calculator(private val logger: ILogger? = null) {
    fun add(a: Double, b: Double): Double = execute("Add", a, b) { a + b }
    fun subtract(a: Double, b: Double): Double = execute("Sub", a, b) { a - b }
    fun multiply(a: Double, b: Double): Double = execute("Mul", a, b) { a * b }
    fun divide(a: Double, b: Double): Double {
        if (b == 0.0) throw IllegalArgumentException("Division by zero is not allowed")
        return execute("Div", a, b) { a / b }
    }

    private fun execute(op: String, a: Double, b: Double, block: () -> Double): Double {
        val res = block()
        logger?.log("$op: $a and $b gives $res")
        return res
    }
}